File: OSCAR/src/gameobjects/Deck.cpp
#include "Deck.h"

Deck::Deck() {
	Deck::initializeDeck();
	srand(time(0));
}

void Deck::initializeDeck() {

	for (int s = 0; s < Deck::NUM_SUITS; s++) {
		for (int v = 2; v < Deck::NUM_VALUES + 2; v++) {
			_fullDeck[s][v - 2] = std::make_unique<Card>(s, v);
		}
	}
}

std::unique_ptr<Card> Deck::takeRandomCard() {

	int suit  = rand() % Deck::NUM_SUITS;
	int value = rand() % Deck::NUM_VALUES;

	// Muy provisional!!!!!!
	if (_fullDeck[suit][value] == nullptr)
		return takeRandomCard();

	// Mark the card as taken
	return std::move( _fullDeck[suit][value]);
}

const void Deck::printAllDeck() const{

	for (int s = 0; s < Deck::NUM_SUITS; s++) {
		for (int v = 0; v < Deck::NUM_VALUES; v++) {
			
			if (_fullDeck[s][v] == nullptr)
				std::cout << Card::missingCardToString();
			else
				std::cout << _fullDeck[s][v]->toString();
		}
	std::cout << '\n';
	}
}

void Deck::retrieveCard(std::unique_ptr<Card>& card) {
	card.swap(_fullDeck[card->getSuit()][card->getValue() - 2]);		// La primera carta(2) se pone en la posicion 0, la 3 en la 1, etc...
}

File: tfgpoker/src/main/java/com/ucm/gameobjects/Deck.java
package com.ucm.gameobjects;

import java.util.Random;

public class Deck {
    private static final int NUM_SUITS = 4; // Ajusta según tu definición
    private static final int NUM_VALUES = 13; // Ajusta según tu definición (2–14 si usas As=14)

    private Card[][] fullDeck;
    private Random random;
    private final String numberOrder = "23456789TJQKA";

    public Deck() {
        fullDeck = new Card[NUM_SUITS][NUM_VALUES];
        random = new Random();
        initializeDeck();
    }

    private void initializeDeck() {
        String suitOrder = "hdcs";
        for (int i = 0; i < NUM_SUITS; i++) {
            for (int j = 2; j < NUM_VALUES + 2; j++) {
                fullDeck[i][j - 2] = new Card(numberOrder.charAt(j), suitOrder.charAt(i));
            }
        }
    }

    public Card takeRandomCard() {
        int suit = random.nextInt(NUM_SUITS);
        int value = random.nextInt(NUM_VALUES);

        if (fullDeck[suit][value] == null) {
            return takeRandomCard(); // vuelve a intentar si ya fue retirada
        }

        Card taken = fullDeck[suit][value];
        fullDeck[suit][value] = null; // la quitamos del mazo
        return taken;
    }

    public void printAllDeck() {
        for (int i = 0; i < NUM_SUITS; i++) {
            for (int j = 0; j < NUM_VALUES; j++) {
                if (fullDeck[i][j] == null) {
                    System.out.print("[xx]");
                } else {
                    System.out.print(fullDeck[i][j].toString());
                }
            }
            System.out.println();
        }
    }

    public void retrieveCard(Card card) {
        // La primera carta (2) se guarda en posición 0, la 3 en la 1, etc.
        int suit = card.getSuit();
        int valueIndex = card.getNumber() - 2;
        fullDeck[suit][valueIndex] = card;
    }
}
